cli/command/container/attach.go

```diff
@@ -65,52 +77,53 @@ func NewAttachCommand(dockerCli command.Cli) *cobra.Command {
 
 	flags := cmd.Flags()
 	flags.BoolVar(&opts.noStdin, "no-stdin", false, "Do not attach STDIN")
+	// Is this feature still used?
+	// It was added in https://github.com/moby/moby/commit/4918769b1ac2d38f23087b766140e6a7f8979310 to allow forwarding signals to containers
+	// Changed in https://github.com/moby/moby/commit/333bc23f21e8423d3085632db99a6d1df227c5f1
+	// And changed to be enabled by default in https://github.com/moby/moby/commit/e0b59ab52b87b8fc15dd5534c3231fdd74843f9f (unless a TTY is attached)
+	// related: https://github.com/moby/moby/commit/e0b59ab52b87b8fc15dd5534c3231fdd74843f9f#commitcomment-25897874
+	// related: https://github.com/moby/moby/issues/9098
+	// related: https://github.com/docker/cli/pull/1841
 	flags.BoolVar(&opts.proxy, "sig-proxy", true, "Proxy all received signals to the process")
 	flags.StringVar(&opts.detachKeys, "detach-keys", "", "Override the key sequence for detaching a container")
 	return cmd
 }
 
-func runAttach(dockerCli command.Cli, opts *attachOptions) error {
-	ctx := context.Background()
+func RunAttach(ctx context.Context, dockerCli command.Cli, containerID string, disableSignalProxy bool, opts container.AttachOptions) error {
 	apiClient := dockerCli.Client()
 
-	// request channel to wait for client
-	resultC, errC := apiClient.ContainerWait(ctx, opts.container, "")
-
-	c, err := inspectContainerAndCheckState(ctx, apiClient, opts.container)
+	c, err := inspectContainerAndCheckState(ctx, apiClient, containerID)
 	if err != nil {
 		return err
 	}
 
-	if err := dockerCli.In().CheckTty(!opts.noStdin, c.Config.Tty); err != nil {
-		return err
-	}
-
-	detachKeys := dockerCli.ConfigFile().DetachKeys
-	if opts.detachKeys != "" {
-		detachKeys = opts.detachKeys
+	if opts.Stdin {
+		if err := dockerCli.In().CheckTty(opts.Stdin, c.Config.Tty); err != nil {
+			return err
+		}
+		if !c.Config.OpenStdin {
+			// TODO(thaJeztah): should this produce an error?
+			opts.Stdin = false
+		}
 	}
 
-	options := container.AttachOptions{
-		Stream:     true,
-		Stdin:      !opts.noStdin && c.Config.OpenStdin,
-		Stdout:     true,
-		Stderr:     true,
-		DetachKeys: detachKeys,
+	if opts.DetachKeys == "" {
+		opts.DetachKeys = dockerCli.ConfigFile().DetachKeys
```

Comment:
Ideally, we'd move this elsewhere so that this function can potentially only depend on the API-client, not the CLI 🤔